### DISCUSSION TITLE
I'm having issues configuring vaultwarden with YubiKey - "Invalid YubiKey OTP provided"

### DISCUSSION THREAD

**Discussion by LordVZlomka** [dani-garcia/vaultwarden#4387] [Q&A] [answered]
**Hi everyone, I encountered an issue setting up my YubiKey 5 NFC.**

The first time I tried setting up my YubiKey, I received an error saying that the YUBICO_CLIENT_ID and YUBICO_SECRET_KEY environment variables needed to be set. 
![Screenshot_2 (1)](https://github.com/dani-garcia/vaultwarden/assets/71541816/bd992564-ce21-4d3e-9415-4439e901b345)

After that, I followed this [tutorial](https://github.com/dani-garcia/vaultwarden/wiki/Enabling-Yubikey-OTP-authentication) and generated the YUBICO_CLIENT_ID and YUBICO_SECRET_KEY. I used the [YubiKey Personalization Tool](https://www.yubico.com/products/services-software/personalization-tools/use/) and then registered my YubiKey with the [OTP server](https://upgrade.yubico.com/getapikey/).

I self-host Vaultwarden via my Synology NAS Docker (container manager). In the "Environment" section of the settings, I've added the YUBICO_CLIENT_ID and YUBICO_SECRET_KEY and their values, which I got earlier.
![Screenshot_3](https://github.com/dani-garcia/vaultwarden/assets/71541816/901ac84c-e815-4269-b9fe-1755986cbfdd)

Now, I can go to the YubiKey setup page (earlier, I couldn't because I was getting an error) when I click "Manage." I selected the YubiKey 1 field, plugged in my YubiKey, and held it for 3 seconds to generate the key. Then I clicked save; it loaded for a bit and gave me an error: "Invalid YubiKey OTP provided."
![Screenshot_2](https://github.com/dani-garcia/vaultwarden/assets/71541816/bfe4ea87-b145-4792-9351-86cc5dfcb093)

**Comment by LordVZlomka** [comment]
UPDATE:

I've contacted YubiKey support, and after some tests, they have confirmed that my YubiKey OTP is working fine, so it's a Vaultwarden issue.
![image](https://github.com/dani-garcia/vaultwarden/assets/71541816/7bbdddb8-7442-4097-91ed-500cddb436cd)

@BlackDex, can you please move this discussion into issues? It seems more appropriate. Since your last message, I didn't want to create a duplicate myself. Thanks.

**Reply by BlackDex** [reply]
It seems to work just fine for me. So it might still be an issue on your side.
Please provide `Support String` which you can generate via the `admin/diagnostics` page.

**Reply by BlackDex** [reply]
Also, what do the logs of Vaultwarden report during that time periode?

**Reply by LordVZlomka** [reply]
Here is the string. How do I view the logs? I couldn't find the setting anywhere in the admin mode.

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.30.3
* Web-vault version: v2024.1.2
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: false
* Internet access via a proxy: false
* DNS Check: false
* Browser/Server Time Check: true
* Server/NTP Time Check: n/a
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.44.0
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** ADMIN_TOKEN


```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*******************************",
  "domain_origin": "*****://*******************************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_change_allowed": true,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "experimental_client_feature_flags": "fido2-vault-credentials",
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": false,
  "push_identity_uri": "https://identity.bitwarden.com",
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": null,
  "smtp_password": null,
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "user_send_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": "97657",
  "yubico_secret_key": "***",
  "yubico_server": null
}
```
</details>

**Reply by BlackDex** [reply]
Your container has no internet access to the outside world.
`Internet access: false`
That is the reason it is failing.

For YubiKey checks we need internet access to validate the key you entered.
We might be able to make the message a bit more descriptive.

But the problem is somewhere on your environment.
Check why the container has not internet access, and once that is solved it should work.

**Reply by LordVZlomka** [reply]
Oh, I see. That makes sense. I run the Vaultwarden via docker on my NAS, and I can access it remotely via a domain I got from Synology, so I'm wondering how it's offline. I gave a link to my friend, and we also successfully accessed it. Am I missing something? Also, I use my mobile device with it, so it's connected.

**Reply by BlackDex** [reply, marked as answer]
That you can go via the internet to the container doesn't mean the container is allowed to the internet it self.
Not sure how those Synology's work and can be configured.

Might be DNS, might be other security/firewall settings. I'm not sure.
But until that internet check reports `true` it will not work.

**Reply by LordVZlomka** [reply]
Yes, it was the firewall. For some reason, it uses ports different from the ones I selected. I couldn't find which one, so I allowed them all for the container's IP adress. It fixed it.

**Reply by LordVZlomka** [reply]
My firewall was blocking UDP 53, and that was the issue.
